iostream_test.go

```diff
@@ -0,0 +1,86 @@
+package iostream_test
+
+import (
+	"bytes"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/PullRequestInc/iostream"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSingleChunkSingleBuffer(t *testing.T) {
+	buf := new(bytes.Buffer)
+	stream := iostream.OpenWriterAtStream(buf, 4, 2)
+
+	written, err := stream.WriteAt([]byte("1234"), 0)
+	stream.Close()
+	require.NoError(t, err)
+	require.Equal(t, 4, written)
+
+	require.NoError(t, err)
+	require.Equal(t, "1234", buf.String())
+}
+
+func TestMultipleChunksSingleBuffer(t *testing.T) {
+	buf := new(bytes.Buffer)
+	stream := iostream.OpenWriterAtStream(buf, 4, 2)
+
+	_, err := stream.WriteAt([]byte("12"), 0)
+	require.NoError(t, err)
+	_, err = stream.WriteAt([]byte("34"), 2)
+	require.NoError(t, err)
+
+	stream.Close()
+
+	require.NoError(t, err)
+	require.Equal(t, "1234", buf.String())
+}
+
+func TestMultipleChunksWithBlockedChunksBeyondBuffer(t *testing.T) {
+	buf := new(bytes.Buffer)
+	stream := iostream.OpenWriterAtStream(buf, 4, 2)
+
+	// this write should get blocked so do it on a separate goroutine
+	go func() {
+		_, err := stream.WriteAt([]byte("90123456"), 8)
+		require.NoError(t, err)
+	}()
+
+	// this write should get blocked so do it on a separate goroutine. it will never
+	// actually get written so we will use it to test behavior of blocked writes getting
+	// unblocked by the Close()
+	cleanedUp := false
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		stream.WriteAt([]byte("far away"), 100)
+		// require.NoError(t, err)
+		cleanedUp = true
+		wg.Done()
```

Comment:
yeah this waitgroup is mostly just being used so that I can verify the blocking writeAt behavior. I might not be following your full question here though, but in general this iostream is just meant for full contents being written to a writerAt so we expected all of the bytes from the beginning through the end to get written unless there is an error that occurs.